hnn_core/pyramidal.py

```diff
@@ -17,9 +17,33 @@
 # Units for gbar: S/cm^2 unless otherwise noted
 
 
+def _read_geo_cells(fname):
+    from neuron import h
+    h.xopen(fname)
+
+    sec_pts = dict()
+    sec_lens = dict()
+    sec_diams = dict()
+    for sec in h.allsec():
+        name = sec.name()
+        sec_diams[name] = getattr(sec, 'diam', None)
+        sec_lens[name] = getattr(sec, 'L', None)
+        sec_pts[name] = list()
+        for idx in range(sec.n3d()):
+            sec_pts[name].append([sec.x3d(idx), sec.y3d(idx), sec.z3d(idx)])
+    return sec_pts, sec_lens, sec_diams
```

Comment:
Since the nrn `h` namespace is global, do we need to worry about creating any objects within the .geo file and not destroying them at the end of this function?

Reply:
yeah, that's why I import neuron locally. Local variables always take precedence in Python ...

Reply:
But `h.allsec()`, for instance, references hoc objects that aren't local. My only concern is for other calls to nrn hoc objects from `NetworkBuilder` or even from calls within distinct pyramidal cell instances. For example, consider the situation where an L2 pyramidal cell is instantiated from a .geo file after the detailed pyramidal cell defined [here](http://cns.iaf.cnrs-gif.fr/files/CELLS/pyramidal264L.geo). If the L2 pyramidal cell has fewer dendrites than the detailed pyramidal cell, the L2 cell's constructor will still end up calling `h.allsec()` which still references all the extra dendrites of the detailed pyramidal cell. Is my understanding wrong?

Reply:
> But h.allsec(), for instance, references hoc objects that aren't local

okay, my thinking was that this `h` is not the same as the global `h`, hence the `allsec` cannot access the hoc objects that are defined in the global `h`. Now, this is only hypothesis at this point. I think the only way to answer the question is to write a tiny 10-15 line script to check what really happens.

> Is my understanding wrong?

I don't know :) Would you be up for writing a tiny script to test what actually happens? Basically do a global import, a local import in a function ... add some sections globally and also locally and see what `h.allsec()` returns. You can share the script on [gist](gist.github.com/)

Reply:
Sure. I thought that at one point we had `neuron.h` imported at many different levels and in many functions throughout hnn_core and ran into issues with them all referencing the same hoc namespace even though no globals were declared in python. An explicit test would be good though.

Reply:
You're right @jasmainak. The potential problem I was suggesting is not a thing. See [here](https://gist.github.com/rythorpe/2022445f44492a79d3d8f9149422dd45). After creating multiple local and global nrn `h.Section` objects and assigning them cable geometry, `h.allsec()` returned objects according to their expected pythonic scope.

Reply:
I think I was getting confused with `h.ParallelContext`.

Reply:
Perfect! Glad to have that sorted out :-D I hope you start to appreciate the power of tiny scripts :)